src/foam/nanos/http/NanoRouter.java

```diff
@@ -64,6 +67,9 @@ protected void service(final HttpServletRequest req, final HttpServletResponse r
 
     resp.setContentType("text/html");
 
+    AppConfig appConfig = (AppConfig) getX().get("appConfig");
+    appConfig.setUrl(((Request) req).getRootURL().toString());
```

Comment:
So when does this occur?
On first url request or on system start?  
At system start the calling url will not be known.

Reply:
For every request received to the servlet, the service() method is called. So each time a new request comes in, a new URL is set.

Reply:
OK, additionally a copy of AppConfig with the calling url must be placed into the subcontext created for each service request. See line 72.  Presently one AppConfig is shared, and callers from different URLs will get last callers URL. 
72: X y = getX().put(HttpServletRequest.class, req)

If AppConfig is a singleton, then two options:
1. make it not a singleton. It may also be possible to make it a multiton using the url as part of key. 
2. change property url to method getUrl(x)

Reply:
This isn't thread-safe since the AppConfig is shared between all threads. AppConfig is a service installed by the NANOS boot process. If you're going to mutate it you should clone() it first.

Reply:
Be we don't have access to Request everywhere we need the url, so perhaps, calculate it, then place it in the context.

Reply:
Request is in the Context so is available anywhere X is.

Reply:
Ok, after a length debate, please do the following:
1. Group - add property url - default to null.
2. Group - add method to create a cloned AppConfig populated with the groups url if provided. 
3. on Login place the cloned AppConfig in the context
4. on Email generation, decorate to add the cloned AppConfig to the context.